ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/BlockAttestationsPool.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.statetransition;
+
+import static tech.pegasys.artemis.datastructures.util.AttestationUtil.getAttesterIndicesIntoCommittee;
+import static tech.pegasys.artemis.datastructures.util.AttestationUtil.isSingleAttester;
+import static tech.pegasys.artemis.datastructures.util.AttestationUtil.setBitsForNewAttestation;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.artemis.datastructures.operations.Attestation;
+import tech.pegasys.artemis.storage.ChainStorage;
+import tech.pegasys.artemis.util.SSZTypes.Bitlist;
+import tech.pegasys.artemis.util.SSZTypes.SSZList;
+import tech.pegasys.artemis.util.config.Constants;
+
+public class BlockAttestationsPool {
+
+  private final ConcurrentHashMap<Bytes32, UnsignedLong> dataRootToSlot = new ConcurrentHashMap<>();
+
+  private final ConcurrentHashMap<Bytes32, Bitlist> unprocessedAttestationsBitlist =
```

Comment:
I think there are visibility issues with these `Bitlist` objects.  Perhaps we could add a thread-safe `Bitlist` implementation that keeps its bits in a concurrent list?

Reply:
Added a ticket: https://pegasys1.atlassian.net/browse/BC-131